module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+
+	logger := e.log.With().Int("worker_id", workerID).Logger()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-e.stop:
+			return
+		case <-e.fetchNotifier.Channel():
+			logger.Debug().Msgf("fetching available requests")
+			e.processAvailableFetchRequests(ctx)
+			logger.Debug().Msgf("done fetching requests")
+		}
+	}
+}
+
+func (e *executionDataRequesterImpl) processAvailableFetchRequests(ctx irrecoverable.SignalerContext) {
+
+	e.log.Debug().Msg("processing fetch requests")
+	for {
+		// retries get priority since the seal height is most likely lower
+		item, ok := e.fetchRetryQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// Check for missing backfills
+
+		item, ok = e.fetchQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// return when queue is empty
+		return
+	}
+}
+
+func (e *executionDataRequesterImpl) processFetchRequest(ctx irrecoverable.SignalerContext, request fetchRequest) error {
+	logger := e.log.With().Str("block_id", request.blockID.String()).Logger()
+
+	logger.Debug().Msgf("processing fetch request for block %d", request.height)
+
+	executionData, err := e.byBlockID(ctx, request.blockID)
+
+	logger.Debug().Msgf("done processing fetch request for block %d", request.height)
+
+	if errors.Is(err, &state_synchronization.MalformedDataError{}) || errors.Is(err, &state_synchronization.BlobSizeLimitExceededError{}) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data id is valid, and not sign the receipt
+		// Crashing isn't a good idea since that would take down access to the network, but this should probably halt the network somehow
+		// maybe we can flag it as bad and not reattempt
+
+		// TODO: add metric
+		logger.Error().Err(err).Msg("HALTING REQUESTER: invalid execution data found")
+		close(e.stop)
+	}
+
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		accepted := e.fetchRetryQueue.Push(request)
+		if !accepted {
+			logger.Warn().Msg("fetch retry queue is full")
+			// TODO: do I need to track this somewhere?
+			// will need a process that periodically re-queues missed requests
+			// possible solution: when retry queue is empty and there is a gap between next notify and last processed, requeue blocks in between
+			return nil
+		}
+
+		e.fetchNotifier.Notify()
+		return nil
+	}
+
+	logger.Debug().Msgf("fetched execution data for height %d", request.height)
+
+	logger.Debug().Msgf("fetch queue has %d requests", e.fetchQueue.Len())
+	logger.Debug().Msgf("fetch retry queue has %d requests", e.fetchRetryQueue.Len())
+
+	accepted := e.cache.Put(request.height, executionData)
+	if !accepted {
+		logger.Warn().Msg("execution data cache is full")
+	}
+
+	e.notificationState.Fetched(request.height, request.blockID)
+	e.notificationNotifier.Notify()
```

Comment:
Here, we could just check if the one we just fetched is the next one we need to notify, and if so, we can directly send that notification, with the pointer to the ExecutionData, to the notifier.

Reply:
this has been refactored in the latest version